### PR TITLE
Client side encryption support for remote storage

### LINKED ISSUE #88: Encryption
Currently, Litestream sends cleartext bytes to a replica such as S3. This works well enough for many use cases as S3 buckets are generally secured through IAM. However, other use cases require additional security measures to prevent attackers from obtaining raw data if a bucket is compromised.

Litestream could implement either encryption to prevent data theft and/or signatures to prevent data alteration.

### PR DESCRIPTION
**Pull request by hifi**
Extremely simple Age wrapper that supports separate recipients and identities. Encrypts both snapshots and WAL files transparently. Technically possible to not have the private key at all if you never restore or do manual restores only.

You can migrate between encrypted/plain by adjusting the identities and recipients respectively but restoring will not work correctly if the remote files aren't encrypted when an identity is configured and vice versa so some care needs to be taken.

Intended use is to encrypt all data before sending out to a potentially untrusted storage provider.

This may conflict with #458 and fixes #88.

### CODE REVIEW

**Comment by benbjohnson** [comment]
@hifi Any objections to upgrading to `age` v1.1.1?

File: cmd/litestream/main.go [hunk 1]
```diff
@@ -323,6 +324,10 @@ type ReplicaConfig struct {
 	User     string `yaml:"user"`
 	Password string `yaml:"password"`
 	KeyPath  string `yaml:"key-path"`
+
+	// Encryption identities and recipients
+	AgeIdentities []string `yaml:"age-identities"`
+	AgeRecipients []string `yaml:"age-recipients"`
```
Comment (benbjohnson) [inline cmd/litestream/main.go]: It'd be nice to reduce the stutter and make the config like:

```yml
age:
  identities: ...
  recipients: ...
```

It can just be an anonymous struct in the config:

```go
type ReplicaConfig struct {
	...
	
	Age struct {
		Identities []string `yaml:"identities"`
		Recipients []string `yaml:"recipients"`
	} `yaml:"age"`
}
```

File: cmd/litestream/main.go [hunk 2]
```diff
@@ -349,6 +354,22 @@ func NewReplicaFromConfig(c *ReplicaConfig, db *litestream.DB) (_ *litestream.Re
 	if v := c.ValidationInterval; v != nil {
 		r.ValidationInterval = *v
 	}
+	for _, str := range c.AgeIdentities {
+		identities, err := age.ParseIdentities(strings.NewReader(str))
+		if err != nil {
+			return nil, err
+		}
+
+		r.AgeIdentities = append(r.AgeIdentities, identities...)
```
Comment (benbjohnson) [inline cmd/litestream/main.go:365]: Why the `append()` instead of just assigning the `identities`? Same goes for the `recipients` below.

File: go.mod [hunk 3]
```diff
@@ -4,14 +4,15 @@ go 1.16
 
 require (
 	cloud.google.com/go/storage v1.15.0
+	filippo.io/age v1.0.0
```
Comment (benbjohnson) [inline go.mod]: Upgrade to latest? `v1.1.1`

File: replica.go [hunk 4]
```diff
@@ -67,6 +68,10 @@ type Replica struct {
 	// If true, replica monitors database for changes automatically.
 	// Set to false if replica is being used synchronously (such as in tests).
 	MonitorEnabled bool
+
+	// Encryption identities and recipients
+	AgeIdentities []age.Identity
+	AgeRecipients []age.Recipient
```
Comment (benbjohnson) [inline replica.go:74]: I don't mind the stutter as much on the struct here. Mostly just the config.

File: replica.go [hunk 5]
```diff
@@ -221,8 +226,20 @@ func (r *Replica) syncWAL(ctx context.Context) (err error) {
...
+	var ew io.WriteCloser = pw
+
+	// Add encryption if we have recipients.
+	if len(r.AgeRecipients) > 0 {
+		var err error
+		ew, err = age.Encrypt(pw, r.AgeRecipients...)
+		defer ew.Close()
+		if err != nil {
+			return err
+		}
```
Comment (benbjohnson) [inline replica.go]: The `defer ew.Close()` needs to be moved after the `if` block. If an error returns then `ew` will be `nil` and panic.

File: replica.go [hunk 6]
```diff
@@ -334,6 +353,15 @@ func (r *Replica) calcPos(ctx context.Context, generation string) (pos Pos, err
 	}
 	defer rd.Close()
 
+	if len(r.AgeIdentities) > 0 {
+		drd, err := age.Decrypt(rd, r.AgeIdentities...)
+		if err != nil {
+			return pos, err
+		}
+
+		rd = ioutil.NopCloser(drd)
```
Comment (benbjohnson) [inline replica.go]: `ioutil` is deprecated. This should be `io.NopCloser()`.

File: replica.go [hunk 7]
```diff
@@ -470,7 +498,23 @@ func (r *Replica) Snapshot(ctx context.Context) (info SnapshotInfo, err error) {
 	// Copy the database file to the LZ4 writer in a separate goroutine.
 	var g errgroup.Group
 	g.Go(func() error {
-		zr := lz4.NewWriter(pw)
+		var wc io.WriteCloser = pw
+
+		// Add encryption if we have recipients.
+		if len(r.AgeRecipients) > 0 {
+			// We need to ensure the pipe is closed.
+			defer pw.Close()
```
Comment (benbjohnson) [inline replica.go]: This seems like it should go at the start of this anonymous function rather than in this `if` block.

File: replica.go [hunk 8]
```diff
@@ -470,7 +498,23 @@ func (r *Replica) Snapshot(ctx context.Context) (info SnapshotInfo, err error) {
...
+
+			var err error
+			wc, err = age.Encrypt(pw, r.AgeRecipients...)
+			defer wc.Close()
+			if err != nil {
+				pw.CloseWithError(err)
+				return err
+			}
```
Comment (benbjohnson) [inline replica.go]: The `defer wc.Close()` should be moved after the `if` block.

**Review by benbjohnson — CHANGES_REQUESTED** [review]

File: cmd/litestream/main.go (hunk 2 above)
Comment (hifi) [reply cmd/litestream/main.go:365]: `age.ParseIdentities` and `age.ParseRecipients` work with NL terminated lists and in the config format I chose to use real YAML lists instead of a multi line string to store them, hence need to loop and append them when parsed one by one.

If preferred I can change both to `string` from `[]string`, then it's just an assignment.

File: go.mod (hunk 3 above)
Comment (hifi) [reply go.mod]: Yeah, when this was originally implemented v1.0.0 was the latest one :+1:

File: replica.go (hunk 7 above)
Comment (hifi) [reply replica.go]: Indeed. It wasn't a problem with the compressor that the pipe wasn't closed but with age the last chunk was cut off so I if guarded it. Shouldn't be a problem it's deferred at the beginning.

File: cmd/litestream/main.go (hunk 2 above)
Comment (benbjohnson) [reply cmd/litestream/main.go:365]: Ah, sorry, I misread it. Yeah, that's fine. 👍

**Review by benbjohnson — APPROVED** [review]

**Comment by benbjohnson** [comment]
Thanks, @hifi! Sorry for the delay.

**Comment by ixxie** [comment]
I would love to use this feature; is it documented anywhere? I can't seem to find it...

**Comment by hifi** [comment]
@ixxie It's not yet in a final release so the website hasn't been updated.

TL;DR for configuration using v0.3.10 beta if my memory serves right:
```yaml
dbs:
  - path: /path/to/database.sqlite3
    replicas:
      - type: file
        path: /path/to/backups
        age:
          identities:
            - <identity>
          recipients:
            - <recipient>
```

For how to generate age identities (private keys) and recipients (public keys), refer to the [age documentation](https://github.com/FiloSottile/age). Public keys are used for restore only and private keys are used for encryption only so it's possible to create a setup where a Litestream instance only encrypts with a public key but cannot restore from the destination on itself without giving it the private key.

**Comment by hifi** [comment]
As far as I know, the releases aren't on a schedule so they'll happen at their own pace.

**Comment by ixxie** [comment]
I see, thanks @hifi!
